add surrogate key blog post (#1877)
fishtown-analytics/docs.getdbt.com: website/blog/2022-08-17-managing-surrogate-keys-in-dbt.md

---

> title: "Managing Surrogate Keys in dbt"

Reviewer: ```suggestion
title: "Managing surrogate keys in dbt"
```

to be consistent with capitalization :)

description: "Wondering how to build a data model with surrogate keys? Dave Connors walks you through two strategies."
slug: managing-surrogate-keys

authors: [dave_connors]

tags: [analytics craft]
hide_table_of_contents: false

> date: 2022-08-17

Reviewer: probably change this tomorrow since it's likely to get merged into tomorrow :)

is_featured: true
---

> ## Introduction

Reviewer: ```suggestion
```
Actually don't think we need this H2, but let me know if you disagree!

> Those who have been building data warehouses for a long time have undoubtedly encountered the challenge of building surrogate keys on their data models. Having a column that uniquely represents each entity helps ensure your data model is complete, does not contain duplicates, and able to join across different data models in your warehouse.  

Reviewer: ```suggestion
Those who have been building <Term id="data-warehouse">data warehouses</Term> for a long time have undoubtedly encountered the challenge of building <Term id="surrogate-key">surrogate keys</Term> on their data models. Having a column that uniquely represents each entity helps ensure your data model is complete, does not contain duplicates, and able to join across different data models in your warehouse.  
```

suggestion: link to glossary pages

Reviewer: I think this should populate them

> Sometimes, we are lucky enough to have data sources with these keys built right in — Shopify data synced via their API, for example, has easy-to-use keys on all the tables written to your warehouse. If this is not the case, or if you build a data model with a compound key (aka the data is unique across multiple dimensions), you will have to rely on some strategy for creating and maintaining these keys yourself. How can you do this with dbt? Let’s dive in:

Reviewer: ```suggestion
Sometimes, we are lucky enough to have data sources with these keys built right in — Shopify data synced via their API, for example, has easy-to-use keys on all the tables written to your warehouse. If this is not the case, or if you build a data model with a compound key (aka the data is unique across multiple dimensions), you will have to rely on some strategy for creating and maintaining these keys yourself. How can you do this with dbt? Let’s dive in.
```

> ### How were surrogate keys managed in the past?

Reviewer: ```suggestion
## How were surrogate keys managed in the past?
```

However, there are also some clear maintenance issues here. Making updates to, say, your products table will require some careful surgical work to ensure the association of cheddar to id 2 is never accidentally changed. You may have heard of the phrase “load your dims before your facts” — this refers to the careful work required to maintain this referential integrity. Additionally, you need to know about the *exact state of the data* before making any updates. This data is **stateful**, making it rigid and more difficult to work with should there be any losses to this data. Imagine trying to rebuild these relationships from scratch!

## MIISKs in dbt

If this is your preferred modeling approach, dbt can absolutely support this workflow! This will likely require you to take advantage of built-in warehouse functionality to generate these MIISKs — in Snowflake, we can use [sequences](https://docs.snowflake.com/en/user-guide/querying-sequences.html), which are objects built exactly for this purpose. We’ll use Snowflake as the example here, but this approach can likely be adapted for other warehouses as well. 

> ### Creating and Maintaining Sequences

Reviewer: ### Creating and maintaining sequences

Idk why won't let me make an in-line comment here

Reviewer: but the suggestion was to change the capitalization to: Creating and maintaining sequences

Author: will update!

    
    | surrogate_key_id | 
    | --- | 
    | 1 | 
    | 2 | 
    | 3 | 
    | 4 | 
    | 5 | x
    | 8 | 
    
    In fact, most cloud platforms [can’t guarantee](https://docs.snowflake.com/en/user-guide/querying-sequences.html#:~:text=Snowflake%20does%20not%20guarantee%20generating%20sequence%20numbers%20without%20gaps.%20The%20generated%20numbers%20consistently%20increase%20in%20value%20(or%20decrease%20in%20value%20if%20the%20step%20size%20is%20negative)%20but%20are%20not%20necessarily%20contiguous) that sequences will be generated without gaps because of their use of parallel processing, even if we *don’t* have a dbt run error — because queries will be spread across multiple compute clusters, each step might query the sequence at different times, which makes it possible to have an out of order sequence result. This is a major consideration in using sequences — if that’s a deal breaker, you may need additional SQL logic in our models (like a `row_number()` function) to guarantee your keys are monotonically increasing. 
    

> - **Views -** Because sequences in Snowflake increment on every query, using them as the surrogate keys for views would mean every time the view is queried, the sequence would increment and therefore change. This strategy would only work for table or incremental models.

Reviewer: ```suggestion
- **<Term id="view">Views</Term> -** Because sequences in Snowflake increment on every query, using them as the surrogate keys for views would mean every time the view is queried, the sequence would increment and therefore change. This strategy would only work for table or incremental models.
```

I'm not sure if this will cause wonkiness with the bold formatting, so recommend playing around with this locally, but suggestion is to link to View glossary page!

- **Ordering -** Since sequences will be regenerated on every run for tables, and every time an incremental model is regenerated, the order of the resulting query determines which records get assigned to each key. In order to maintain referential integrity (i.e. product_id 1 always means mozzarella), you need to build in `ORDER BY` statements to your models. This can cause adverse performance during table builds.
- **“Load your dims before your facts” -** This strategy can also lead to some very messy DAGs in order to keep relationships intact in your project. As mentioned above, it’s imperative that each product record results in the same surrogate key value every time dbt is run. Additionally, this means that any table that needs to read from this table needs to run downstream of that initial process. This can lead to bottlenecks at runtime.

Even though configuring MIISKs with sequences can be pretty well automated, it’s a bit of a brittle process that relies on a lot of assumptions and requires a whole lot of bandwidth from the data team to recreate the warehouse should anything go haywire. 

> ## Hashed keys

Reviewer: ```suggestion
## Hashed surrogate keys
```

```sql
# in models/reports/daily_user_orders.sql
with 

orders as (
	select * from {{ ref('fct_orders') }} 
),

agg as (

	select 
		date_trunc(day, order_date) as report_date 	
		user_id, 
		count(*) as total_orders
  from orders
  group by 1,2

),

final as (
	
	select
		{{ dbt_utils.surrogate_key([
				'report_date', 
				'user_id'
			])
		}} as unique_key, 
		*
	from agg

)

select * from final
```

> Using hashed keys makes our transformations idempotent — every dbt run results in the same exact outputs every time. I can safely delete all my non-source objects in my warehouse, execute a dbt run and be right back where I started (though I wouldn’t necessarily recommend this 😅).  

Reviewer: ```suggestion
Using hashed keys makes our transformations <Term id="idempotent" /> — every dbt run results in the same exact outputs every time. I can safely delete all my non-source objects in my warehouse, execute a dbt run and be right back where I started (though I wouldn’t necessarily recommend this 😅).  
```

The analytical warehouses we use now no longer have the same constraints that traditional warehouses had  — joins on strings aren’t notably less performant than those on integers, and storing slightly larger values in the surrogate key column is peanuts given the relative cost of storage on these platforms. This strategy also removes the need for tight coupling of transformations to propagate the surrogate key values across our project — anywhere the inputs for the surrogate keys are present, the hashing function produces the same keys, so we can take advantage of parallel processing in our warehouse and avoid the bottlenecks we had before. 

### Caveats

This strategy is not without its caveats either!

> - **Collisions -** Although it is exceedingly rare, it is possible for two different sets of inputs to produce the same outputs, causing erroneous duplicate records in your dataset. Using an MD5 hash (the default for the `dbt_utils.surrogate_key` macro), you have a 50% of a collision when you get up to 2^64 records (1.84 x 10E19 aka a whole lot of data). While very unlikely, it’s certainly something to consider for truly massive datasets.

Reviewer: ```suggestion
- **Collisions -** Although it's *exceedingly* rare, depending on the hashing algorithm you use, it's possible for two different sets of inputs to produce the same outputs, causing erroneous duplicate records in your dataset. Using an MD5 hash (the default for the `dbt_utils.surrogate_key` macro), you have a 50% of a collision when you get up to 2^64 records (1.84 x 10E19 aka a whole lot of data). While [very very very unlikely](https://docs.getdbt.com/terms/surrogate-key#a-note-on-hashing-algorithms), it’s certainly something to consider for truly massive datasets.
```

> - **Performance -** As mentioned above, hashed keys result in long string-type values. On truly massive datasets, this can cause some performance issues. Unlike MIISKs, string values can’t be easily partitioned to improve query performance.

Reviewer: This section is somewhat ominous haha. Do you have recommendations for huge datasets using hashed surrogate keys? Ex. Using incremental/table materializations.

Author: for performance, i've sometimes see using different datatypes a from different hashing functions as the right move, as described in the next bullet point! I think i can add a sentence here to make that clearer.

Author: the collisions piece i think is mostly unavoidable unfortunately 😬

Reviewer: Oh sorry should have been more specific! The section on Performance seems a little ominous to me!!

Reviewer: I think you could call out using other hashing algorithms (SHA256 or something) in the Collisions section, but again, an md5 collision is so very unlikely so it could be overkill

Author: I rearranged them and softened the language in performance a bit! let me know what you think!

Reviewer: Looking great to me :)
